### PR TITLE
Enhancements in CUDA Support

### PR DESCRIPTION
**Pull request by ar90n**
This PR introduces several updates to improve CUDA support which is derived from #461

Added specifications for a new option to provide users with more flexibility.

1. Incorporated support for newer CUDA and CuDNN versions.
1. Implemented libcudnn8-dev installation support
1. Implemented nvcc installation support
1. Implemented CUDA toolkit installation support.

These enhancements aim to make the developer experience smoother for those working on CUDA projects.

### CODE REVIEW

**Comment by ar90n** [comment]
@microsoft-github-policy-service agree

File: src/nvidia-cuda/install.sh [hunk 1]
```diff
@@ -55,23 +58,33 @@ fi
 echo "Installing CUDA libraries..."
 apt-get install -yq "$cuda_pkg"
 
-if [ "$INSTALL_CUDNN" = "true" ]; then
+if [ "$INSTALL_CUDNN" = "true" ] || [ "$INSTALL_CUDNNDEV" = "true" ]; then
     # Ensure that the requested version of cuDNN is available AND compatible
     cudnn_pkg_version="libcudnn8=${CUDNN_VERSION}-1+cuda${CUDA_VERSION}"
     if ! apt-cache show "$cudnn_pkg_version"; then
         echo "The requested version of cuDNN is not available: cuDNN $CUDNN_VERSION for CUDA $CUDA_VERSION"
```
Comment (samruddhikhandale) [inline src/nvidia-cuda/install.sh:65]: Let's say `INSTALL_CUDNN` is false, but `INSTALL_CUDNNDEV` is true, and there's case of unavailability, then the Feature will fail with this error. This seems completely unexpected, instead, can we move changes for installing cuDNN DEV to an isolated condition?

File: test/nvidia-cuda/install_all_options.sh [hunk 2]
```diff
@@ -8,8 +8,14 @@ source dev-container-features-test-lib
 # Check installation of libcudnn8
 check "libcudnn.so.8" test 1 -eq "$(find /usr -name 'libcudnn.so.8' | wc -l)"
 
+# Check installation of libcudnn8-dev
```
Comment (samruddhikhandale) [inline test/nvidia-cuda/install_all_options.sh:11]: I don't expect these checks to be pass as the Feature options are `false` by default, you would need to update the following scenario 👇  

https://github.com/devcontainers/features/blob/50a7b160f0c7ea1956e7897f5843ef57332e6141/test/nvidia-cuda/scenarios.json#L2-L10

Also, you'd need to rename the scenario ^ name to `install_all_options`

File: src/nvidia-cuda/devcontainer-feature.json [hunk 3]
```diff
@@ -10,14 +10,27 @@
       "default": false,
       "description": "Additionally install CUDA Deep Neural Network (cuDNN) shared library"
```
Comment (samruddhikhandale) [inline src/nvidia-cuda/devcontainer-feature.json:11]: Can we bump the minor version? Thanks! 

https://github.com/devcontainers/features/blob/50a7b160f0c7ea1956e7897f5843ef57332e6141/src/nvidia-cuda/devcontainer-feature.json#L3

File: src/nvidia-cuda/README.md [hunk 4]
```diff
@@ -16,7 +16,10 @@ Installs shared libraries for NVIDIA CUDA.
 | Options Id | Description | Type | Default Value |
 |-----|-----|-----|-----|
 | installCudnn | Additionally install CUDA Deep Neural Network (cuDNN) shared library | boolean | false |
+| installCudnnDev | Additionally install CUDA Deep Neural Network (cuDNN) development libraries and headers | boolean | false |
 | installNvtx | Additionally install NVIDIA Tools Extension (NVTX) | boolean | false |
+| installNvcc | Additionally install NVIDIA CUDA Compiler (NVCC) | boolean | false |
```
Comment (samruddhikhandale) [inline src/nvidia-cuda/README.md:21]: I don't see `installNvcc` added to this PR, wondering if this is intentional or if we are missing some code.

**Review by samruddhikhandale — COMMENTED** [review]
Thank you for adding cool new Feature options, we appreciate it.
Left some comments!

File: test/nvidia-cuda/install_all_options.sh (hunk 2 above)
Comment (samruddhikhandale) [reply test/nvidia-cuda/install_all_options.sh:11]: Ref: https://github.com/devcontainers/features/actions/runs/6311978556/job/17158820859?pr=699#step:4:20

File: src/nvidia-cuda/README.md [hunk 5]
```diff
@@ -16,7 +16,10 @@ Installs shared libraries for NVIDIA CUDA.
 | Options Id | Description | Type | Default Value |
 |-----|-----|-----|-----|
 | installCudnn | Additionally install CUDA Deep Neural Network (cuDNN) shared library | boolean | false |
+| installCudnnDev | Additionally install CUDA Deep Neural Network (cuDNN) development libraries and headers | boolean | false |
```
[the remark this answers is not in the document]
Comment (ar90n) [reply src/nvidia-cuda/README.md]: In https://github.com/devcontainers/features/pull/699/commits/3ec682aaf94c700adb6901cde46fbe2077a75253, I reverted the modifications of README.md

File: test/nvidia-cuda/install_all_options.sh (hunk 2 above)
Comment (ar90n) [reply test/nvidia-cuda/install_all_options.sh:11]: In https://github.com/devcontainers/features/pull/699/commits/58ae521c84a715d1856bbe05473f621e56015c68, I fixed scenarios.json to pass this test.

File: src/nvidia-cuda/devcontainer-feature.json (hunk 3 above)
Comment (ar90n) [reply src/nvidia-cuda/devcontainer-feature.json:11]: I used 1.1.0 as the new version. Is this correct? I should use 1.0.8?

File: src/nvidia-cuda/README.md (hunk 4 above)
Comment (ar90n) [reply src/nvidia-cuda/README.md]: This is my mistake. I just remembered that I need to remove this line. CUDA toolkit contains nvcc. So we don't have to install nvcc explicitly.

File: src/nvidia-cuda/install.sh (hunk 1 above)
Comment (ar90n) [reply src/nvidia-cuda/install.sh]: I made the installation of cudnn-dev isolated.

**Comment by ar90n** [comment]
@samruddhikhandale 
Thanks for your thorough review and detailed comments. I fixed some issues you mentioned. Please check them.

**Review by samruddhikhandale — APPROVED** [review]
Looks good to me, opened https://github.com/devcontainers/features/issues/704 for the failing test. 

(This PR is not causing the failures, hence, we are good to merge this in)

**Pull request merged**